### DISCUSSION TITLE
Safe conversion of monorepo to pnpm workspace

### DISCUSSION THREAD

**Discussion by jfirebaugh** [pnpm/pnpm#6156] [Q&A]
Suppose I have a monorepo with subdirectories containing independent `package.json`/`pnpm-lock.yaml` pairs, that does not yet use [workspaces](https://pnpm.io/workspaces), and I want to begin using a workspace. Is there a way do this that preserves the precise version locks that exist pre-workspace?

In my testing, pnpm does not preserve the semantics of the existing `pnpm-lock.yaml` files when converting a project to use a workspace. For instance, if I have two directories, `p1` and `p2`, both with:

```
  "dependencies": {
    "repeat-string": "^1.5.2"
  }
```

in `package.json`, but locked at (say) 1.5.2 in p1 and 1.6.1 in p2, then after adding a `pnpm-workspace.yaml`

```
packages:
  - p1
  - p2
```

and running `pnpm i`, the generated root-level `pnpm-lock.yaml` locks both p1 and p2 to the _same_ version of `repeat-string`. The behavior I would expect is that pnpm generates a root-level lockfile that preserves the locking defined by the nested `pnpm-lock.yaml`s.

I would be reluctant to convert a large project to workspaces without this guarantee, as despite everyone's best efforts, we know that even semver-compatible upgrades sometimes introduce bugs and incompatibilities (not to mention the supply chain risks associated with unconstrained upgrades). That's why lockfiles exist in the first place.

**Comment by iobuhov** [comment]
I'm not sure is this what you looking for, but there [shared-workspace-lockfile](https://pnpm.io/npmrc#shared-workspace-lockfile) setting that, as I understand, when set to false will not use root lockfile but will create lock file per project. BTW, never used it, so, I might be wrong.

**Reply by jfirebaugh** [reply]
Thanks for that suggestion! This does seem to preserve the exact version dependencies as desired, but in doing so gives up many of the benefits of workspaces as well, such as sharing a single copy of dependencies that can be shared (e.g. where the same version _is_ used across multiple packages).

**Comment by jfirebaugh** [comment]
I looked at what npm does, and it also does not preserve existing version locks when converting to a workspace. The equivalent feature request for npm is https://github.com/npm/rfcs/issues/521.
